### PR TITLE
Update routing readme and script to set up TLS

### PR DESCRIPTION
**Pull request by liggitt**

### CODE REVIEW

**Comment by liggitt** [comment]
@pweil- review

File: hack/install-router.sh [hunk 1]
```diff
@@ -1,34 +1,71 @@
 #!/bin/bash
```
Comment (smarterclayton) [inline hack/install-router.sh]: @pweil- this has reached way over complexity - we need to figure out how you install routers in a saner way (admittedly, the bulk of this is secrets, but this is painful for end users too).  @csrwng this is the sort of thing that secrets has to solve for us.

Comment (liggitt) [inline hack/install-router.sh]: I kept wanting a template, but that requires talking to the server, which was previously optional here, and I tried to preserve. What I really want is a security context for the secrets, and a service telling me the address and root certs of the master. If wishes were fishes...

Comment (smarterclayton) [inline hack/install-router.sh]: Hopefully soon...

Comment (pweil-) [inline hack/install-router.sh]: Hmmm, yes, this has grown complicated with the introduction of the secure master.  Needs to be easier.

@liggitt thanks for putting this together!  I'll give it a full read through and test in the morning.  Nothing jumps out at me with a quick glance.

File: docs/routing.md [hunk 2]
```diff
@@ -30,8 +30,10 @@ Once it is pulled it will start and be visible in the `docker ps` list of contai
     [vagrant@openshiftdev origin]$ make clean && make
     [vagrant@openshiftdev origin]$ export PATH=/data/src/github.com/openshift/origin/_output/local/bin/linux/amd64:$PATH
     [vagrant@openshiftdev origin]$ sudo /data/src/github.com/openshift/origin/_output/local/bin/linux/amd64/openshift start &
-    [vagrant@openshiftdev origin]$ hack/install-router.sh {router_id} {master ip}
-    [vagrant@openshiftdev origin]$ openshift kubectl get pods
+    [vagrant@openshiftdev origin]$ sudo chmod a+r /data/src/github.com/openshift/origin/openshift.local.certificates/admin/*
+    [vagrant@openshiftdev origin]$ export KUBECONFIG=/data/src/github.com/openshift/origin/openshift.local.certificates/admin/.kubeconfig
+    [vagrant@openshiftdev origin]$ hack/install-router.sh {router_id} {master_url}
+    [vagrant@openshiftdev origin]$ openshift cli get pods
```
Comment (pweil-) [inline docs/routing.md]: This no longer reflects the way the router would need to be started

File: hack/install-router.sh [hunk 3]
```diff
@@ -1,34 +1,71 @@
...
 fi
 
-# update the template file
-cp ${OS_ROOT}/images/router/haproxy/pod.json /tmp/router.json
-sed -i s/MASTER_IP/${MASTER_IP}/ /tmp/router.json
-sed -i s/ROUTER_ID/${ROUTER_ID}/g /tmp/router.json
+OPENSHIFT_INSECURE="${OPENSHIFT_INSECURE:-false}"
+OPENSHIFT_CA_DATA="${OPENSHIFT_CA_DATA:-}"
+OPENSHIFT_CERT_DATA="${OPENSHIFT_CERT_DATA:-}"
+OPENSHIFT_KEY_DATA="${OPENSHIFT_KEY_DATA:-}"
```
Comment (pweil-) [inline hack/install-router.sh]: I tried setting these locally and starting the router with this message from the container:

`[vagrant@openshiftdev origin]$ docker logs 6e54
F0123 14:32:45.451188       1 router.go:46] Unable to configure Kubernetes client: crypto/tls: failed to parse certificate PEM data`

**Comment by pweil-** [comment]
Ok, think I found the issue.  The `pod.json` that `hack/install-router.sh` is using doesn't contain the env vars necessary.  Also, the installer is expecting the contents of the file and not the file name as I expected.

The pod should change to look like 
```
{
    "kind": "Pod",
    "apiVersion": "v1beta1",
    "id": "ROUTER_ID",
    "desiredState": {
        "manifest": {
            "version": "v1beta2",
            "containers": [
                {
                    "name": "origin-haproxy-router-ROUTER_ID",
                    "image": "openshift/origin-haproxy-router",
                    "ports": [
                        {
                            "containerPort": 80,
                            "hostPort": 80
                        },
                        {
                            "containerPort": 443,
                            "hostPort": 443
                        }
                    ],
                    "env": [
                        {
                            "name": "OPENSHIFT_MASTER",
                            "value": "${OPENSHIFT_MASTER}"
                        },
                        {
                            "name": "OPENSHIFT_CA_DATA",
                            "value": "${OPENSHIFT_CA_DATA}"
                        },
                        {
                            "name": "OPENSHIFT_CERT_DATA",
                            "value": "${OPENSHIFT_CERT_DATA}"
                        },
                        {
                            "name": "OPENSHIFT_KEY_DATA",
                            "value": "${OPENSHIFT_KEY_DATA}"
                        }
                    ],
                    "command": ["--loglevel=4"],
                    "imagePullPolicy": "PullIfNotPresent"
                }
            ],
            "restartPolicy": {
                "always": {}
            }
        }
    }
}
```
And the installer can take a cert dir (or all the individual files, but just the cert dir seems easier for now) and then pull the contents like (and format like you had it):

```
if [[ "${MASTER_URL}" == "https"* ]]; then
	if [[ "$OPENSHIFT_CA_DATA" == "" ]]; then
		echo "Running against an HTTPS master (${MASTER_URL}) without a trusted certificate bundle."
		echo "Set \$OPENSHIFT_CA_DATA to the contents of the root certificate bundle to start securely next time."
		echo "Starting insecurely..."
		OPENSHIFT_INSECURE=true
	fi

	# TODO: remove this once secrets are available
	# Escape cert data for json
	CERT_DIR=/data/src/github.com/openshift/origin/openshift.local.certificates
	OPENSHIFT_CA_DATA=$(<$CERT_DIR/openshift-client/root.crt)
	OPENSHIFT_CERT_DATA=$(<$CERT_DIR/openshift-client/cert.crt)
	OPENSHIFT_KEY_DATA=$(<$CERT_DIR/openshift-client/key.key)


	OPENSHIFT_CA_DATA="${OPENSHIFT_CA_DATA//$'\n'/\\\\n}"
	OPENSHIFT_CERT_DATA="${OPENSHIFT_CERT_DATA//$'\n'/\\\\n}"
	OPENSHIFT_KEY_DATA="${OPENSHIFT_KEY_DATA//$'\n'/\\\\n}"
else
	OPENSHIFT_INSECURE=""
	OPENSHIFT_CA_DATA=""
	OPENSHIFT_CERT_DATA=""
	OPENSHIFT_KEY_DATA=""
fi
```

that seems to work fine in my vagrant environment

**Comment by pweil-** [comment]
@liggitt @csrwng does ^ make sense?

**Comment by pweil-** [comment]
Actually, after all this, I think my problem was....me.  I see the correct items in the pod.json, I expected the installer to take file names.  My only issue then is that the docs don't accurately show how to use the installer anymore.  I'll retest just for my sanity since I know how it is supposed to work now.

@csrwng @liggitt disregard the other ramblings

**Comment by pweil-** [comment]
Ok, retest worked fine.  The only thing I noticed is on a clean, single machine vagrant environment the installer will not work with the generated certificates because they are signed by an unknown authority.  I had to add `--certificate-authority=/data/src/github.com/openshift/origin/openshift.local.certificates/openshift-client/root.crt`.  

So

1.  The doc doesn't reflect the certificate contents in the env vars for a secure start
1.  The installer doesn't work with the OOTB certs without modification

**Comment by liggitt** [comment]
@pweil- fixed the docs... which installer are you talking about?

**Comment by pweil-** [comment]
`hack/install-router.sh` .  Here is what I saw:

```
[vagrant@openshiftdev origin]$ hack/install-router.sh router https://10.0.2.15:8443
Creating router file and starting pod...
F0123 11:19:04.211542    1712 create.go:75] Post https://localhost:8443/api/v1beta1/pods?namespace=default: x509: certificate signed by unknown authority
```

**Comment by liggitt** [comment]
when I made the certs readable, set OPENSHIFT_CA_DATA and KUBECONFIG as per the doc, it created and ran without incident for me... what is your KUBECONFIG env?

**Comment by liggitt** [comment]
also, can you just run `osc get pods`?

**Comment by pweil-** [comment]
derp, completely missed the KUBECONFIG step on my retest.  Thanks for suffering through my comments.  Doc changes look good to me.  Tested in vagrant successfully.

LGTM

**Comment by pweil-** [comment]
@smarterclayton when this is merged I need to rebase https://github.com/openshift/origin/pull/652 and I *believe* it's ready for more review.  Bring the pain! :hammer:

**Comment by liggitt** [comment]
[merge]

**Comment by liggitt** [comment]
re[merge]

**Pull request merged**